src/main/java/com/google/devtools/build/lib/rules/java/JavaInfoBuildHelper.java

```diff
@@ -284,6 +285,22 @@ public JavaInfo createJavaCompileAction(
 
     JavaToolchainProvider toolchainProvider = javaToolchain;
 
+    JavaPluginInfo pluginInfo = mergeExportedJavaPluginInfo(plugins, deps);
+    Builder<String> allJavacOptsBuilder = ImmutableList.<String>builder()
+        .addAll(toolchainProvider.getJavacOptions(starlarkRuleContext.getRuleContext()))
+        .addAll(javaSemantics.getCompatibleJavacOptions(starlarkRuleContext.getRuleContext(),
+            toolchainProvider));
+    if (pluginInfo.plugins().processorClasses().toList()
```

Comment:
@cushon asks:

> Are we OK with flattening the nested set here?
>
> Another option that preserves this behaviour would be to be to add the javacopt in JavaBuilder instead (I think an earlier version of the PR did that). But maybe these nested sets are small enough it isn't a big concern, up to you.

Reply:
`processorClasses().toSet()` is called in a few places that I believe are executed during the analysis phase, so I assume it already ends up flattened.

I pushed a commit to use `toSet` rather than `toList` for consistency.